internal/command/postgres/restart.go

```diff
@@ -52,136 +57,282 @@ func runRestart(ctx context.Context) error {
 		return fmt.Errorf("app %s is not a Postgres app", app.Name)
 	}
 
+	agentclient, err := agent.Establish(ctx, client)
+	if err != nil {
+		return fmt.Errorf("can't establish agent %w", err)
+	}
+
+	dialer, err := agentclient.Dialer(ctx, app.Organization.Slug)
+	if err != nil {
+		return fmt.Errorf("can't build tunnel for %s: %s", app.Organization.Slug, err)
+	}
+	ctx = agent.DialerWithContext(ctx, dialer)
+
 	switch app.PlatformVersion {
 	case "nomad":
-		if err := hasRequiredVersionOnNomad(app, MinPostgresHaVersion, MinPostgresHaVersion); err != nil {
+		if err = hasRequiredVersionOnNomad(app, MinPostgresHaVersion, MinPostgresHaVersion); err != nil {
 			return err
 		}
-		return restartNomadCluster(ctx, app)
-	case "machines":
-		agentclient, err := agent.Establish(ctx, client)
+		vms, err := client.GetAllocations(ctx, app.Name, false)
 		if err != nil {
-			return fmt.Errorf("can't establish agent %w", err)
+			return fmt.Errorf("can't fetch allocations: %w", err)
 		}
+		if flag.GetBool(ctx, "hard") {
+			return nomadHardRestart(ctx, vms)
+		}
+		return nomadSoftRestart(ctx, vms)
 
-		dialer, err := agentclient.Dialer(ctx, app.Organization.Slug)
+	case "machines":
+		flapsClient, err := flaps.New(ctx, app)
 		if err != nil {
-			return fmt.Errorf("can't build tunnel for %s: %s", app.Organization.Slug, err)
+			return fmt.Errorf("list of machines could not be retrieved: %w", err)
 		}
+		ctx = flaps.NewContext(ctx, flapsClient)
 
-		leader, err := fetchLeader(ctx, app, dialer)
+		machines, err := flapsClient.List(ctx, "started")
 		if err != nil {
-			return fmt.Errorf("can't fetch leader: %w", err)
+			return fmt.Errorf("machines could not be retrieved %w", err)
+		}
+		if len(machines) == 0 {
+			return fmt.Errorf("no machines found")
 		}
-		if err := hasRequiredVersionOnMachines(leader, MinPostgresHaVersion, MinPostgresHaVersion); err != nil {
+		if err := hasRequiredVersionOnMachines(machines[0], MinPostgresHaVersion, MinPostgresHaVersion); err != nil {
 			return err
 		}
-		return restartMachinesCluster(ctx, app)
+		if flag.GetBool(ctx, "hard") {
+			return machinesHardRestart(ctx, machines)
+		}
+		return machinesSoftRestart(ctx, machines)
 	}
 
 	return nil
 }
 
-func restartMachinesCluster(ctx context.Context, app *api.AppCompact) error {
+func machinesSoftRestart(ctx context.Context, machines []*api.Machine) error {
```

Comment:
We should test this, but I think performing a failover here would actually result in more downtime.  Process level restarts are super fast, so we probably don't need to worry about performing a failover here.